src/main/java/org/voltdb/exportclient/KinesisFirehoseExportClient.java

```diff
@@ -0,0 +1,264 @@
+            Record record = new Record();
+            try {
+                final ExportRowData row = decodeRow(rowData);
+                if (!writeRow(row.values, csv, m_skipInternals, m_binaryEncoding, m_ODBCDateformat.get())) {
+                    return false;
+                }
+                csv.flush();
+
+                String data = stringer.toString();
+                record.withData(ByteBuffer.wrap(data.getBytes()));
+            } catch(IOException e) {
+                rateLimitedLogError(m_logger, "Failed to build record: %s", Throwables.getStackTraceAsString(e));
+                throw new RestartBlockException(true);
+            } finally {
+                try { csv.close(); } catch (IOException e) {}
+            }
+
+            m_records.add(record);
+            return true;
+        }
+
+        @Override
+        public void sourceNoLongerAdvertised(AdvertisedDataSource source)
+        {
+            m_es.shutdown();
+            try {
+                m_es.awaitTermination(365, TimeUnit.DAYS);
+            } catch (InterruptedException e) {
+                Throwables.propagate(e);
+            }
+        }
+
+        @Override
+        public void onBlockStart() throws RestartBlockException
+        {
+            m_records = new ArrayList<Record>();
+        }
+
+        @Override
+        public void onBlockCompletion() throws RestartBlockException
+        {
+            try {
+                int recordsSize = m_records.size();
+                List<Record> recordsList;
+                int sleepTime = 0;
+                while (recordsSize > 0) {
+                    if (sleepTime > 0)
+                        Thread.sleep(sleepTime);
+                    // PutRecordBatchRequest can not contain more than 500 records
+                    if (recordsSize > 500) {
+                        recordsList = m_records.subList(0, 499);
+                        m_records = new ArrayList<Record>(m_records.subList(500, recordsSize-1));
+                        recordsSize = m_records.size();
+                    } else {
+                        recordsList = new ArrayList<Record>(m_records);
+                        m_records.clear();
+                        recordsSize = 0;
+                    }
+                    PutRecordBatchRequest batchRequest = new PutRecordBatchRequest().
+                            withDeliveryStreamName(m_streamName).
+                            withRecords(recordsList);
+                    PutRecordBatchResult res = m_firehoseClient.putRecordBatch(batchRequest);
+                    if (res.getFailedPutCount() > 0) {
+                        for (PutRecordBatchResponseEntry entry : res.getRequestResponses()) {
+                            if (entry.getErrorMessage() != null && !entry.getErrorMessage().contains("Slow down.")) {
+                                rateLimitedLogError(m_logger, "Record failed with response: %s", entry.getErrorMessage());
```

Comment:
Is this going to export some records more than once when the block contains more than 500 records?